nucleus/common/glassfish-api/src/main/java/org/glassfish/api/invocation/InvocationManagerImpl.java

```diff
@@ -68,347 +71,334 @@
 @Singleton
 public class InvocationManagerImpl implements InvocationManager {
 
-    // This TLS variable stores an ArrayList.
-    // The ArrayList contains ComponentInvocation objects which represent
-    // the stack of invocations on this thread. Accesses to the ArrayList
-    // don't need to be synchronized because each thread has its own ArrayList.
-    private InheritableThreadLocal<InvocationArray<ComponentInvocation>> frames;
-
-    private final ThreadLocal<Stack<ApplicationEnvironment>> applicationEnvironments = withInitial(Stack<ApplicationEnvironment>::new);
```

Comment:
NB. Replacing `Stack` here is more of a modernisation to avoid `sychronized` in favour of CAS based synchronisation.